### **Name** : Yogesh Kumar Gopal
### **Student Id**: 8996403

2. Use the iris flower dataset from the sklearn library and cross_val_score against the logistic regression model to measure the performance of each. 

Record your observations/reflections in the Python notebook.

# Install libraries, if not installed (Optional)

In [29]:
%pip install pandas scikit-learn numpy

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.0 -> 25.1.1
[notice] To update, run: python.exe -m pip install --upgrade pip


# Add the imports

In [30]:
import pandas as pd
import numpy as np
from sklearn.datasets import load_iris
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score

# Load the dataset

In [31]:
# Load the iris dataset
the_iris = load_iris()
# Features
XFeatures = the_iris.data  
# Targets
yTarget = the_iris.target 

print("\n Loading the isis dataset")
print("Iris Dataset loaded successfully. Shape of X:", XFeatures.shape, "Shape of y:", yTarget.shape)
print("\nFirst 5 rows of features (X):")
print(pd.DataFrame(XFeatures, columns=the_iris.feature_names).head(5))
print("\nTarget names:", the_iris.target_names)


 Loading the isis dataset
Iris Dataset loaded successfully. Shape of X: (150, 4) Shape of y: (150,)

First 5 rows of features (X):
   sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

Target names: ['setosa' 'versicolor' 'virginica']


# Initialize the Logistic Regression model.

In [32]:
# We'll set 'solver' to 'liblinear' and 'multi_class' to 'ovr' explicitly for clarity, though 'lbfgs' is often default.
# 'max_iter' is increased to ensure convergence for this dataset.
logistic_model = LogisticRegression(solver='liblinear', multi_class='ovr', max_iter=200)

In [33]:
print("Logistic Regression model initialized.")
print(f"Model parameters: {logistic_model.get_params()}")

Logistic Regression model initialized.
Model parameters: {'C': 1.0, 'class_weight': None, 'dual': False, 'fit_intercept': True, 'intercept_scaling': 1, 'l1_ratio': None, 'max_iter': 200, 'multi_class': 'ovr', 'n_jobs': None, 'penalty': 'l2', 'random_state': None, 'solver': 'liblinear', 'tol': 0.0001, 'verbose': 0, 'warm_start': False}


# Perform Cross-Validation

In [34]:
print("\nPerforming Cross-Validation")
# Perform cross-validation with 5 folds
# Cross-validation scores will give us an array of accuracy scores for each fold.
scores = cross_val_score(logistic_model, XFeatures, yTarget, cv=5, scoring='accuracy')
print(f"Cross-validation scores for each fold (cv=5): {scores}")


Performing Cross-Validation
Cross-validation scores for each fold (cv=5): [1.         0.96666667 0.93333333 0.9        1.        ]


c:\Users\yogeshkumar\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\linear_model\_logistic.py:1256: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. Use OneVsRestClassifier(LogisticRegression(..)) instead. Leave it to its default value to avoid this warning.
  warnings.warn(
c:\Users\yogeshkumar\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\linear_model\_logistic.py:1256: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. Use OneVsRestClassifier(LogisticRegression(..)) instead. Leave it to its default value to avoid this warning.
  warnings.warn(
c:\Users\yogeshkumar\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\linear_model\_logistic.py:1256: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. Use OneVsRestClassifier(LogisticRegression(..)) instead. Leave it to its default value to avoid this warning.
  warnings.warn(
c

# Analyze Results

In [35]:
print("\nAnalyzing Cross-Validation Results")
m_accuracy = np.mean(scores)
s_accuracy = np.std(scores)

print(f"Mean Accuracy: {m_accuracy:.4f}")
print(f"Standard Deviation of Accuracy: {s_accuracy:.4f}")


Analyzing Cross-Validation Results
Mean Accuracy: 0.9600
Standard Deviation of Accuracy: 0.0389


# Record Observations/Reflections

**The Iris Dataset: An Ideal Testbed**
The Iris dataset is a classic choice for demonstrating machine learning principles. It's concise and well-structured, making it perfectly suited for illustrating concepts like cross-validation in a straightforward manner.

**Logistic Regression's Strong Performance**
Logistic Regression proved to be highly effective on the Iris dataset, even with its inherent linear decision boundaries. This strong performance is due to the dataset's nature: the Iris classes are largely linearly separable. Specifically, 'setosa' is quite distinct from the others, and 'versicolor' and 'virginica' also show good, albeit less pronounced, separation.

**Insights from Cross-Validation Scores**
The cross_val_score function provided individual accuracy scores for each fold of the cross-validation. These scores are crucial because they reflect the model's performance when trained and evaluated on diverse subsets of the data, offering a more comprehensive view than a single test.

**Mean Accuracy: A Robust Estimate**
The average accuracy across all folds was [Insert mean_accuracy value here, e.g., 0.9667]. This mean score offers a far more reliable estimate of the model's true performance on unseen data. Unlike a single train-test split, which can be swayed by the specific data points in its test set, the mean accuracy from cross-validation provides a more generalized understanding.

**Standard Deviation: Indicating Model Consistency**
The standard deviation of the scores, [Insert std_accuracy value here, e.g., 0.0211], tells us about the variability of the model's performance. A low standard deviation, as seen here, signifies that the model's performance is remarkably consistent regardless of how the data is partitioned into folds. This consistency is a strong indicator of good generalization capabilities.

**The Value of Cross-Validation**
This exercise clearly highlights the benefits of cross-validation. It provides a more robust estimate of a model's performance and is an excellent tool for detecting potential overfitting. If a model performs exceptionally well on one specific data split but poorly on others (leading to a high standard deviation), it's a red flag for overfitting. In this case, the low standard deviation suggests that our Logistic Regression model is quite stable and avoids this pitfall.

**Model Stability and Generalization**
Ultimately, the combination of a high mean accuracy and a low standard deviation strongly indicates that Logistic Regression is both a stable and effective classifier for the Iris dataset. It demonstrates excellent generalization, meaning it can reliably classify new, unseen Iris samples.
